_posts/2019-01-21-api-best-practices.md

```diff
@@ -0,0 +1,90 @@
+---
+
+If you've been a web developer for any length of time at all these days, you've no doubt used at least a
+few web based APIs here and there.  It's possible that you've even written one (or more!) yourself.
+API design is a rich topic with a lot of deep roots, but some cursory research will show a number of best
+practices that public facing APIs should implement.  Understanding these practices will give you a firm basis
+for judging wether you've been saddled with using a poorly designed API, and allow you to design more useful
+APIs when it's your turn.
+
+<!--more-->
+
+Over the years, I've used a *lot* of APIs myself, and here are some best practices that I see getting short shrift:
+
+###  Versioning
+Versioning an API is something that's surprisingly difficult to do well, and yet absolutely necessary.  As time goes
+by and your API gains more and more users, the time will come when you need to make changes to the public interface
+that may break a lot of your users' existing code.  Versioning allows you to neatly sidestep this problem by just
+putting the breaking changes in a new version.  Actually doing this, however, will need a [lot](https://restfulapi.net/versioning/) [of](https://apigee.com/about/blog/developer/common-misconceptions-about-api-versioning) [careful](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7) [thought](https://blog.restcase.com/restful-api-versioning-insights/).  What's most important here isn't necessarily the way you choose to do it, however.  What's most important is that you *bother to do it at all*.
+
+### Return navigation information in link headers, where applicable
+If you're returning paginated data, or results that are otherwise part of a large data set, it would be a good
+idea to return [link headers](https://tools.ietf.org/html/rfc5988).  These headers let you include navigation
+links, along with type information describing their use.  Including them makes your users' lives easier, and aides
+in the creation of full featured navigation UIs.
+
+### If your API is rate limited, return the current limit counters
+Otherwise, you'll be forcing the developers using your API to implement their own counting logic to track the status.
+They probably won't like that.  As a bonus, if you ever have to change the rate limiting policy on your API, it now
+becomes a breaking change for all of your users.
+
+Avoid that headache and do your users a favor by returning rate limiting counters in the response to every rate
+limited request.  The headers you'll usually want to use are:
+
+**X-Rate-Limit-Limit** - The number of requests allowed in a given period  
+**X-Rate-Limit-Remaining** - The number of requests left in the current period  
+**X-Rate-Limit-Reset** - Seconds until the current period elapses (may not be useful if you implement
+rolling rate limit windows)  
+
+And while we're discussing rate limiting, you may also want to consider returning the appropriate error codes
```

Comment:
Maybe you can link `error codes` to something like https://en.wikipedia.org/wiki/List_of_HTTP_status_codes